File: Leetcode/784.cpp
/**
 * 784. 字母大小写全排列
 * link：https://leetcode.cn/problems/letter-case-permutation/
 */

class Solution
{
public:
    vector<string> letterCasePermutation(string s)
    {
        vector<string> result;
        vector<char> current(s.size(), 0);
        dfs(result, current, s, 0);
        return result;
    }

    void dfs(vector<string> &result, vector<char> &current, const string &src, int index)
    {
        while (index < src.size() && !isalpha(src[index]))
        {
            current[index] = src[index];
            index++;
        }
        if (index >= src.size())
        {
            result.emplace_back(current.begin(), current.end());
            return;
        }
        current[index] = tolower(src[index]);
        dfs(result, current, src, index + 1);

        current[index] = toupper(src[index]);
        dfs(result, current, src, index + 1);
    }
};

File: Leetcode/784.js
/**
 * 784. 字母大小写全排列
 * link：https://leetcode.cn/problems/letter-case-permutation/
 */

/**
 * @param {string} s
 * @return {string[]}
 */
var letterCasePermutation = function (s) {
    const isAlpha = function (ch) {
        const code = ch.toLowerCase().charCodeAt(0);
        return (code >= 'a'.charCodeAt(0) && code <= 'z'.charCodeAt(0))
    }

    let result = [];
    const n = s.length;
    let currentResult = new Array(n);
    const dfs = function (index) {
        while (index < n && !isAlpha(s[index])) {
            currentResult[index] = s[index];
            index++;
        }
        if (index >= n) {
            result.push(currentResult.join(''));
            return;
        }
        currentResult[index] = s[index].toLowerCase();
        dfs(index + 1);

        currentResult[index] = s[index].toUpperCase();
        dfs(index + 1);
    }

    dfs(0);
    return result;
};